### PR TITLE
Match query support for semantic_text fields - POC DO NOT MERGE

### PR DESCRIPTION
**Pull request by Mikep86**
POC implementation of `match` query support for `semantic_text` fields. The technical goal of this POC is to update the `match` query rewrite logic such that it is rewritten to a `semantic` query when we detect that we are querying a `semantic_text` field.

There were two big challenges implementing this POC:

- Rewriting to a `semantic` query *must* happen on the coordinating node, where mappings are not available
- `match` query code lives in `server`, while `semantic` query code lives in the `inference` plugin

I solved these challenges in the POC by adding the name of the query that should be used to query the inference field to `InferenceFieldMetadata`. Since this is stored in index metadata, it is available on the coordinating node. I also added `QueryBuilderService`, which stores query builder creation functions for query builders that can be created using a field name and query string. `QueryBuilderService` is then populated using the `getQueryBuilders` plugin hook. I added `QueryBuilderService` to `QueryRewriteContext`, making it available to the `match` query during rewrite on the coordinating node. The query name from `InferenceFieldMetadata` is then used to create the proper query builder using `QueryBuilderService`.

Because of scoring incompatibility issues, the `match` query can be used to query `semantic_text` fields only when that is the only field type being queried. That is to say, you can't query a `text` field and `semantic_text` field using the same `match` query.

Putting it all together, it means you can now do things like this:

```json
PUT _inference/sparse_embedding/my-elser-endpoint
{
  "service": "elser",
  "service_settings": {
    "num_allocations": 1,
    "num_threads": 1
  },
  "task_settings": {}
}

PUT my-index-1
{
    "mappings": {
        "properties": {
            "inference_field": {
                "type": "semantic_text",
                "inference_id": "my-elser-endpoint"
            },
            "text_field": {
                "type": "text"
            }
        }
    }
}

PUT my-index-2
{
    "mappings": {
        "properties": {
            "inference_field": {
                "type": "text"
            },
            "text_field": {
                "type": "text"
            }
        }
    }
}

POST my-index-1/_doc/1
{
  "inference_field": "a test value",
  "text_field": "a test value"
}

POST my-index-2/_doc/1
{
  "inference_field": "a test value",
  "text_field": "a test value"
}

GET my-index-1/_search
{
  "query": {
    "match": {
      "inference_field": "test"
    }
  }
}

GET my-index-1/_search
{
  "query": {
    "match": {
      "text_field": "test"
    }
  }
}

// Fails due to combination of text & semantic_text fields
GET my-index-*/_search
{
  "query": {
    "match": {
      "inference_field": "test"
    }
  }
}
```

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/QueryBuilderServiceBuilder.java
Comment (Mikep86) [inline server/src/main/java/org/elasticsearch/index/query/QueryBuilderServiceBuilder.java:1]: We can probably rewrite this builder as a static `build` method in `QueryBuilderService`

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java [hunk 1]
```diff
@@ -368,6 +372,45 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
...
+                Collection<IndexMetadata> indexMetadataCollection = resolvedIndices.getConcreteLocalIndicesMetadata().values();
+
+                boolean foundNonInferenceField = false;
+                String inferenceFieldQueryName = null;
+                for (IndexMetadata indexMetadata : indexMetadataCollection) {
+                    InferenceFieldMetadata inferenceFieldMetadata = indexMetadata.getInferenceFields().get(fieldName);
+                    if (inferenceFieldMetadata != null) {
+                        if (inferenceFieldQueryName != null
+                            && inferenceFieldMetadata.getQueryName().equals(inferenceFieldQueryName) == false) {
+                            throw new IllegalArgumentException("Detected incompatible inference field queries");
```
Comment (Mikep86) [inline server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:391]: These error messages leave much to be desired. Will iterate on them to make them more informative.

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java [hunk 2]
```diff
@@ -368,6 +372,45 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
...
+                        }
+
+                        inferenceFieldQueryName = inferenceFieldMetadata.getQueryName();
+                    } else {
+                        foundNonInferenceField = true;
+                    }
+                }
+
+                if (foundNonInferenceField && inferenceFieldQueryName != null) {
+                    throw new IllegalArgumentException("Cannot query inference fields and non-inference fields at the same time");
```
Comment (kderusso) [inline server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: Would it be better to dis max this and just return everything, knowing there will be better scores for `semantic_text` fields, than straight up erroring?

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java [hunk 3]
```diff
@@ -368,6 +372,45 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
...
+                    InferenceFieldMetadata inferenceFieldMetadata = indexMetadata.getInferenceFields().get(fieldName);
+                    if (inferenceFieldMetadata != null) {
+                        if (inferenceFieldQueryName != null
+                            && inferenceFieldMetadata.getQueryName().equals(inferenceFieldQueryName) == false) {
+                            throw new IllegalArgumentException("Detected incompatible inference field queries");
+                        }
+
+                        inferenceFieldQueryName = inferenceFieldMetadata.getQueryName();
+                    } else {
+                        foundNonInferenceField = true;
```
Comment (kderusso) [inline server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:396]: Can probably break out of the loop here once we get a single `true` value.

**Review by kderusso — COMMENTED** [review]
I like this approach - it's clean overall. 

Some early feedback: 

- I wonder if we should err on the side of being more permissive, rather than throwing an error for a lot of these queries. Maybe if we're trying to `match` on multiple query types it's OK, wrap in a `dis_max` query or use RRF or semantic reranking as a recommendation?

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 2 above)
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: The issue is actually the opposite: When querying a `semantic_text` field that uses a dense vector model, all the scores will be in the [0.0-1.0] range. BM25 scores are essentially guaranteed to be greater than this, so all the `semantic_text` results would be at the end of the result set.

Comment (kderusso) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: Ah, interesting, I ran through your script and

```
GET my-index-1/_search
{
  "query": {
    "match": {
      "inference_field": "test"
    }
  }
}
```

returned a score of 12.725867, while

```
GET my-index-1/_search
{
  "query": {
    "match": {
      "text_field": "test"
    }
  }
}
```
returned a score of 0.2876821.

File: server/src/main/java/org/elasticsearch/index/query/QueryBuilderServiceBuilder.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+public class QueryBuilderServiceBuilder {
+    private PluginsService pluginsService;
+
+    public QueryBuilderServiceBuilder pluginsService(PluginsService pluginsService) {
+        this.pluginsService = pluginsService;
+        return this;
+    }
+
+    public QueryBuilderService build() {
+        Objects.requireNonNull(pluginsService);
```
[the remark this answers is not in the document]
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/query/QueryBuilderServiceBuilder.java:28]: I think we can remove the builder entirely actually and just have a static `build` method in `QueryBuilderService` instead

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 2 above)
Comment (kderusso) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: Ah, nevermind, you noted dense vector not sparse vector - that makes sense. However, I think the point of either using dis max or just allowing it is potentially better than failing fast in this case.

**Comment by Mikep86** [comment]
@kderusso 

> Maybe if we're trying to match on multiple query types it's OK, wrap in a dis_max query or use RRF or semantic reranking as a recommendation?

Wrapping in a `dis_max` query doesn't work due to score range differences described [in this comment](https://github.com/elastic/elasticsearch/pull/112166#discussion_r1729437302). We could update the error messages to suggest RRF or reranking to combine the result sets, that gives the user a thread for a solution at least.

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 2 above)
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: That's just one example though. We know that BM25 scores are unbounded and combining them naively with scores in a bounded range is going to result in bad relevance. IMO we shouldn't do something we know will result in bad relevance.

**Comment by kderusso** [comment]
> Wrapping in a dis_max query doesn't work due to score range differences described https://github.com/elastic/elasticsearch/pull/112166#discussion_r1729437302. We could update the error messages to suggest RRF or reranking to combine the result sets, that gives the user a thread for a solution at least.

Right - I get that, I think we should document RRF or reranking as a recommended solution but isn't returning matching results better than an error here?

**Comment by Mikep86** [comment]
> Right - I get that, I think we should document RRF or reranking as a recommended solution but isn't returning matching results better than an error here?

Not necessarily. The error at least tells the user what they are doing wrong. If we return results, how are they to know that anything is wrong? All they will know is that they get bad results out of the box, giving them a bad impression of Elasticsearch.

**Review by carlosdelest — COMMENTED** [review]
This looks good - rewriting to a semantic query in the coordinator makes sense.

I have some thoughts for simplifying this, at least initially. LMKWYT!

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 2 above)
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: Good callout about the different sorting use case. I think we could potentially handle this by adding sort info to the rewrite context so we can detect this case and allow it.

I don't think documentation is a full solution though. So many users don't even read the documentation, and those who do skim over entire critical sections all the time. So adding a section saying essentially "WARNING: if you use a query pattern like this, your relevance will be silently broken" isn't a complete solution to me. We already have problems putting too much cognitive load on our customers, and particularly in that light this seems egregious.

I think documentation can be part of a larger solution though. For example, we could try following:

- Emit a warning header when we detect a scenario where relevance is likely compromised
- Block such scenarios by default, but add a param to allow them at the user's request

In addition to documenting the issue.

Basically, my position is that we need to be proactive about alerting the user that what they are doing is likely to compromise relevance, and we need to do so in a way that does not require reading the documentation.

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java [hunk 5]
```diff
@@ -368,6 +372,45 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
         builder.endObject();
     }
 
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        QueryBuilder rewritten = super.doRewrite(queryRewriteContext);
+
+        if (rewritten == this && queryRewriteContext.getClass() == QueryRewriteContext.class) {
```
[the remark this answers is not in the document]
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:379]: I think we can refactor to avoid the `getClass()` comparison. This was the quick & dirty approach.

And just to clarify, this rewrite phase does not run in the can_match phase. It runs just before the can_match phase. It may seem like a pedantic detail, but it's important because the can_match phase is skipped for most search requests.

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java [hunk 6]
```diff
@@ -368,6 +372,45 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
...
+                        foundNonInferenceField = true;
+                    }
+                }
+
+                if (foundNonInferenceField && inferenceFieldQueryName != null) {
+                    throw new IllegalArgumentException("Cannot query inference fields and non-inference fields at the same time");
+                }
+
+                if (inferenceFieldQueryName != null) {
+                    rewritten = queryRewriteContext.getQueryBuilderService()
```
[the remark this answers is not in the document]
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:405]: That's part of the motivation, but I also added `QueryBuilderService` in an attempt to add something that can help with use cases outside of semantic search. There is nothing specific to semantic search in the `QueryBuilderService` implementation; it can be used to make a query implemented in any plugin accessible in `server` or any other plugin without major refactoring.

I agree we can probably simplify this POC implementation. You are correct that I wrote this with the idea in mind that someday we will have more than one inference field and inference query type. Before simplifying though, I think we should go through a thought experiment of how we will be forward-compatible if/when we add more inference field types or inference query types.

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:379]: > And just to clarify, this rewrite phase does not run in the can_match phase. It runs just before the can_match phase. It may seem like a pedantic detail, but it's important because the can_match phase is skipped for most search requests.


Understood - my comment referred to the current `CoordinatorRewriteContext`, which is really what runs on the can_match phase instead of actually being what is used in the coordinator.

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 2 above)
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: I like the warning approach 👍 

I understand your reasons, but I still think we shouldn't block the user from doing  something that has some legitimate use cases. 

We can decide later on this (separate PR for future work) and/or open it to a broader group for feedback.

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:405]: > We could even move semanticQuery() method from the SemanticTextFieldMapper to the InferenceFieldMapper interface to signal the intent.

That approach won't work because we don't have access to the mappings on the coordinating node. All we have access to is `InferenceFieldMetadata`, which is why I added the query name to that.

Let's discuss offline about how we can simplify. I agree there is room to do so, but it may not be as straightforward as it appears.

**Comment by mayya-sharipova** [comment]
@Mikep86 I have not studied the PR deeply, so forgive my naive questions:  but why do we need to do a rewrite to a semantic query on a coordinating node?  Why can't it be done this rewrite we are already on a shard and have all mappings?

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java [hunk 7]
```diff
@@ -368,6 +372,45 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
         builder.endObject();
     }
 
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
```
[the remark this answers is not in the document]
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:376]: `doCoordinatorRewrite` is specifically for rewriting queries against data streams on the coordinator node. It's very poorly named and confuses a lot of people. It's on my wish list to refactor the rewrite contexts to fix things like this :)

In it's current form, this rewrite logic will be called once: on the coordinator node, just before the can-match phase (if it is active).

File: server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java [hunk 8]
```diff
@@ -32,28 +33,38 @@
...
     private static final String SOURCE_FIELDS_FIELD = "source_fields";
+    private static final String QUERY_NAME_FIELD = "query_name";
 
     private final String name;
     private final String inferenceId;
     private final String[] sourceFields;
+    private final String queryName;
 
-    public InferenceFieldMetadata(String name, String inferenceId, String[] sourceFields) {
+    public InferenceFieldMetadata(String name, String inferenceId, String[] sourceFields, String queryName) {
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java:43]: Why are we adding this new metadata at all?

**Comment by Mikep86** [comment]
@mayya-sharipova 

> why do we need to do a rewrite to a semantic query on a coordinating node? Why this rewrite can't it be done when we are already on a shard and have all mappings? Is it because we want to make only a single inference per request?

We must rewrite on the coordinating node because the semantic query performs inference on the coordinating node. If we were to wait and rewrite on the shards, we would have to perform inference N times (where N is the number of shards).

> Why do we need a queryName parameter for InferenceFieldMetadata(String name, String inferenceId, String[] sourceFields, String queryName)? Could there be any different value from semantic?

Currently, no. This POC is written to be forward-compatible with any inference queries we may add in the future. @carlosdelest and I discussed (at a high level) how we could simplify in [this comment thread](https://github.com/elastic/elasticsearch/pull/112166#discussion_r1738389352). We should continue that discussion offline.

File: server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java (hunk 8 above)
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java:43]: For forward-compatibility with any potential new queries against inference fields. It's probably overkill, but I wanted to write the POC with that potential in mind. We can simplify for the production implementation.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java:43]: > For forward-compatibility with any potential new queries against inference fields. It's probably overkill, but I wanted to write the POC with that potential in mind. We can simplify for the production implementation.

OK, this doesn't seem necessary and was strange to me. It would only be required for the edge case of a mixed cluster environment & us trying to allow "new code" to be used on an "old node". This over complicates things and we shouldn't do this.

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 2 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: I am strongly in favour of NOT throwing an error and using the default scores from different types of queries.  Currently it works this way: for example if a user runs match query on a numeric field in one index (which always produces scores of `1` ) and on a text field in another index, docs with their scores will be combined without errors.

File: server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java (hunk 8 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java:43]: I agree, it seems overkill for the current need.

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:376]: Thanks for the clarification.

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 2 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: I agree with @mayya-sharipova . We shouldn't fail here. Its a horrible experience. While its weird to have things scored in different distributions, this is technically already occurring with BM25 over two indices with vastly different term statistics.

Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: Good to see a consensus forming here. Given that the best dense vector scores will almost always be less than 
the worst BM25 scores, I think we should at least alert the user to the scoring mismatch in a more proactive way than just documentation. What does everyone think about a warning header in this scenario?

Comment (kderusso) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: I think a warning header might be overkill here.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:401]: Agreed, no warning headers, no logging, no nothing. We just need to ensure that `explain` works.

**Comment by Mikep86** [comment]
Closing this PR as it has served its purpose for collecting feedback on the POC
